## Source acquisition and file classification

- This layer code should be framework-agnostic and should be run it before Docling/LangChain.
- Docling is strong for downstream parsing because it supports many formats and preserves text, tables, pictures, hierarchy, layout boxes, and provenance in a unified document model.
- OCRmyPDF is a good OCR fallback because it adds searchable text layers to scanned PDFs and is tolerant of born-digital PDFs that do not need OCR.
- LangChain can then consume the cleaned outputs later through loaders such as DoclingLoader, Unstructured, JSONLoader, BSHTMLLoader, and CSVLoader.

## PDF layout and reading-order extraction



**Can docling handle multi col text in correct order?**

In [33]:
from pprint import pprint
from pathlib import Path
from docling.document_converter import DocumentConverter

In [34]:
data_folder_path = Path("/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named")
file_path = data_folder_path / "2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text.pdf"
file_path

PosixPath('/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named/2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text.pdf')

In [35]:
converter = DocumentConverter()
doc = converter.convert(file_path).document

[INFO] 2026-04-15 15:14:48,212 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 15:14:48,437 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 15:14:48,439 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 15:14:48,853 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 15:14:48,979 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer.onnx
[INFO] 2026-04-15 15:14:48,980 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer

In [36]:
doc.export_to_markdown()

"<!-- image -->\n\n## N E W S L E T T E R\n\nV h i s d p  %  R o p h i f   1 / 1 0\n\nRspdWl PdrsnpWshnm sn :cuWmbd        Rntsg SteW Pdnodmr        CdBgWladWt Onmcr Opnidbs        Dhdkc RdlhmWpr\n\nBNUDQ OGNSN BNTQSDRW NE C9UHC I- FTADQMHBJ&amp; P.CMANTROCPCS BNI\n\nMNQ9 KHUHMFRSNM\n\ns I write this note it feels like spring is on its way. The sun's a little warmer and the birds a little more vocal. It feels like we're turning a corner in this pandemic too. And there's good news about Mono Basin stream restoration.\n\nAll we have to do is wait for spring. But unlike the changing season, almost everything else-certainly everything in this issue of the Newsletter -happens because of human eff  orts. Sometimes those eff  orts are monumental; more often they're incremental and together, they add up.\n\nThat's certainly the case with stream restoration. We have made incremental eff  orts for years against DWP's deliberate delay tactics. Persistent phone calls, letters, and meetings are fi 

In [37]:
type(doc)

docling_core.types.doc.document.DoclingDocument

In [38]:
docdict = dict(doc)
docdict

{'schema_name': 'DoclingDocument',
 'version': '1.10.0',
 'name': '2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text',
 'origin': DocumentOrigin(mimetype='application/pdf', binary_hash=7132794411961773591, filename='2021-winter-spring-mono-lake-newsletter-medium-mix-multi-col-text.pdf', uri=None),
 'furniture': GroupItem(self_ref='#/furniture', parent=None, children=[], content_layer=<ContentLayer.FURNITURE: 'furniture'>, meta=None, name='_root_', label=<GroupLabel.UNSPECIFIED: 'unspecified'>),
 'body': GroupItem(self_ref='#/body', parent=None, children=[RefItem(cref='#/pictures/0'), RefItem(cref='#/texts/2'), RefItem(cref='#/texts/3'), RefItem(cref='#/texts/4'), RefItem(cref='#/texts/5'), RefItem(cref='#/texts/6'), RefItem(cref='#/texts/7'), RefItem(cref='#/texts/8'), RefItem(cref='#/texts/9'), RefItem(cref='#/texts/10'), RefItem(cref='#/texts/11'), RefItem(cref='#/texts/12'), RefItem(cref='#/texts/13'), RefItem(cref='#/texts/14'), RefItem(cref='#/pictures/1'), RefItem

In [39]:
docdict.keys()

dict_keys(['schema_name', 'version', 'name', 'origin', 'furniture', 'body', 'groups', 'texts', 'pictures', 'tables', 'key_value_items', 'form_items', 'field_regions', 'field_items', 'pages'])

In [40]:
type(docdict['pages']), docdict['pages']

(dict,
 {1: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=1),
  2: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=2),
  3: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=3),
  4: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=4),
  5: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5),
  6: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=6),
  7: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=7),
  8: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=8),
  9: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=9),
  10: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=10),
  11: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=11),
  12: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=12),
  13: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=13

In [41]:
docdict['pages'][3]

PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=3)

**How can I print content of a specific page?**
- 'pages' contains a dict which is a list of PageItem objects, there is no page content, it also says image=None
- maybe 'texts' will contain those info

In [42]:
docdict['texts']

[TextItem(self_ref='#/texts/0', parent=RefItem(cref='#/pictures/0'), children=[], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.TEXT: 'text'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=85.50278399999999, t=695.9567095955999, r=265.999018848, b=629.9933517955529, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 4))], source=[], comments=[], orig='LNMN', text='LNMN', formatting=None, hyperlink=None),
 TextItem(self_ref='#/texts/1', parent=RefItem(cref='#/pictures/0'), children=[], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.TEXT: 'text'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=340.666992, t=695.9567095955999, r=524.890242576, b=629.9933517955529, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 4))], source=[], comments=[], orig='K IA', text='K IA', formatting=None, hyperlink=None),
 SectionHeaderItem(self_ref='#/texts/2', parent=RefItem(cref='#/body'), children=[], content_

In [43]:
len(docdict['texts']), type(docdict['texts'])

(695, list)

- There is a list of 695 TextItem s
- Let's observe the item at idx 0

In [44]:
dict(docdict['texts'][0])

{'self_ref': '#/texts/0',
 'parent': RefItem(cref='#/pictures/0'),
 'children': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'meta': None,
 'label': <DocItemLabel.TEXT: 'text'>,
 'prov': [ProvenanceItem(page_no=1, bbox=BoundingBox(l=85.50278399999999, t=695.9567095955999, r=265.999018848, b=629.9933517955529, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 4))],
 'source': [],
 'comments': [],
 'orig': 'LNMN',
 'text': 'LNMN',
 'formatting': None,
 'hyperlink': None}

- Now, let's observe all the text items from page no 3

In [45]:
for textitem in docdict['texts']:
    textitem = dict(textitem)
    if textitem['prov'][0].page_no == 3:
        print(f"{textitem['self_ref']} -----------")
        pprint(textitem)
        print()

#/texts/38 -----------
{'children': [],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'formatting': None,
 'hyperlink': None,
 'label': <DocItemLabel.SECTION_HEADER: 'section_header'>,
 'level': 1,
 'meta': None,
 'orig': 'RspdWl pdrsnpWshnm onhrdc sn kdWo enpvWpc hm 1-10',
 'parent': RefItem(cref='#/body'),
 'prov': [ProvenanceItem(page_no=3, bbox=BoundingBox(l=133.858992, t=742.231919202, r=478.142775552, b=682.862469940578, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 49))],
 'self_ref': '#/texts/38',
 'source': [],
 'text': 'RspdWl pdrsnpWshnm onhrdc sn kdWo enpvWpc hm 1-10'}

#/texts/39 -----------
{'children': [],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'formatting': None,
 'hyperlink': None,
 'label': <DocItemLabel.TEXT: 'text'>,
 'meta': None,
 'orig': 'Nudpbnlhmf CSO cdkZx bkdZpr oZsg sn Zbshnm ax Edneepdx LbPthkjhm',
 'parent': RefItem(cref='#/body'),
 'prov': [ProvenanceItem(page_no=3, bbox=BoundingBox(l=194.0

**Observations**
- Multicolumn and multiblocks of texts
- Among all text item blocks 'DocItemLabel.TEXT' are extracted correctly
- But 'DocItemLabel.CAPTION', 'DocItemLabel.PAGE_FOOTER' or 'DocItemLabel.SECTION_HEADER' are not extracted correctly

Let's check some other document:

In [46]:
data_folder_path = Path("/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named")
file_path = data_folder_path / "nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf"
file_path

PosixPath('/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named/nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf')

In [47]:
converter = DocumentConverter()
doc = converter.convert(file_path).document

[INFO] 2026-04-15 15:50:39,980 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 15:50:40,033 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 15:50:40,034 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 15:50:40,213 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 15:50:40,254 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer.onnx
[INFO] 2026-04-15 15:50:40,254 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer

In [48]:
doc.export_to_markdown()

'## NATURE ALBERTA\n\n<!-- image -->\n\nA COMMUNITY CONNECTED BY A LOVE OF NATURE\n\nMeet Alberta\'s Hummingbirds\n\nCoal: Policy, Pushback, and the Price We Pay\n\nMissing Mountain Meadows\n\nMAGAZINE\n\nHinton\'s Botanical Treasure Trove\n\n<!-- image -->\n\n## It\'s an Alberta thing.\n\nHasse Lake - Lake Aeration Conservation Site (D3-206)\n\nThe Alberta Discover Guide is a FREE annual publication, website, and app that features over 790 conservation sites across Alberta that outdoor enthusiasts can access for hunting, fishing, hiking, foraging, and photography.\n\nalbertadiscoverguide.com\n\n<!-- image -->\n\n<!-- image -->\n\nphoto: ACA, Meagan Butler\n\nArt generated by Doug Pedersen\n\n## CONTENTS\n\n## SPRING 2026\n\n- 2 President\'s Message\n- 3 Nature Alberta News\n- 4 Caribou Update: Plans for Extirpation\n- 6 A Hummingbird Primer\n- 11 Of Tree Rings and Tea Leaves\n- 14 Botanical Treasures of Hinton\'s Beaver Boardwalk\n- 18 Coal Confusion: Throwing Dust in the Air\n- 23 Bo

In [49]:
type(doc)

docling_core.types.doc.document.DoclingDocument

In [50]:
docdict = dict(doc)
docdict

{'schema_name': 'DoclingDocument',
 'version': '1.10.0',
 'name': 'nature-alberta-spring-2026-magazine-medium-mix-multi-col-text',
 'origin': DocumentOrigin(mimetype='application/pdf', binary_hash=9295684724258981057, filename='nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf', uri=None),
 'furniture': GroupItem(self_ref='#/furniture', parent=None, children=[], content_layer=<ContentLayer.FURNITURE: 'furniture'>, meta=None, name='_root_', label=<GroupLabel.UNSPECIFIED: 'unspecified'>),
 'body': GroupItem(self_ref='#/body', parent=None, children=[RefItem(cref='#/texts/0'), RefItem(cref='#/pictures/0'), RefItem(cref='#/texts/5'), RefItem(cref='#/texts/6'), RefItem(cref='#/texts/7'), RefItem(cref='#/texts/8'), RefItem(cref='#/texts/9'), RefItem(cref='#/texts/10'), RefItem(cref='#/texts/11'), RefItem(cref='#/pictures/1'), RefItem(cref='#/texts/14'), RefItem(cref='#/texts/15'), RefItem(cref='#/texts/16'), RefItem(cref='#/texts/17'), RefItem(cref='#/pictures/2'), RefItem(cre

In [51]:
docdict.keys()

dict_keys(['schema_name', 'version', 'name', 'origin', 'furniture', 'body', 'groups', 'texts', 'pictures', 'tables', 'key_value_items', 'form_items', 'field_regions', 'field_items', 'pages'])

In [52]:
type(docdict['pages']), docdict['pages']

(dict,
 {1: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=1),
  2: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=2),
  3: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=3),
  4: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=4),
  5: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5),
  6: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=6),
  7: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=7),
  8: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=8),
  9: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=9),
  10: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=10),
  11: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=11),
  12: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=12),
  13: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=13

In [53]:
docdict['pages'][5]

PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5)

In [54]:
docdict['texts']

[SectionHeaderItem(self_ref='#/texts/0', parent=RefItem(cref='#/body'), children=[], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.SECTION_HEADER: 'section_header'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=33.440749999999994, t=760.3216, r=478.46648500000003, b=575.515137359263, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 14))], source=[], comments=[], orig='NATURE ALBERTA', text='NATURE ALBERTA', formatting=None, hyperlink=None, level=1),
 TextItem(self_ref='#/texts/1', parent=RefItem(cref='#/pictures/0'), children=[], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.TEXT: 'text'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=507.2007, t=751.5801, r=566.6396999999998, b=742.8161510805502, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 11))], source=[], comments=[], orig='SPRING 2026', text='SPRING 2026', formatting=None, hyperlink=None),
 TextItem(self_ref='#/texts/2', pare

In [55]:
len(docdict['texts']), type(docdict['texts'])

(744, list)

In [56]:
dict(docdict['texts'][0])

{'self_ref': '#/texts/0',
 'parent': RefItem(cref='#/body'),
 'children': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'meta': None,
 'label': <DocItemLabel.SECTION_HEADER: 'section_header'>,
 'prov': [ProvenanceItem(page_no=1, bbox=BoundingBox(l=33.440749999999994, t=760.3216, r=478.46648500000003, b=575.515137359263, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 14))],
 'source': [],
 'comments': [],
 'orig': 'NATURE ALBERTA',
 'text': 'NATURE ALBERTA',
 'formatting': None,
 'hyperlink': None,
 'level': 1}

In [58]:
for textitem in docdict['texts']:
    textitem = dict(textitem)
    if textitem['prov'][0].page_no == 5:
        print(f"{textitem['self_ref']} -----------")
        pprint(textitem)
        print()

#/texts/97 -----------
{'children': [],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'formatting': None,
 'hyperlink': None,
 'label': <DocItemLabel.SECTION_HEADER: 'section_header'>,
 'level': 1,
 'meta': None,
 'orig': 'Nature Alberta News',
 'parent': RefItem(cref='#/body'),
 'prov': [ProvenanceItem(page_no=5, bbox=BoundingBox(l=52.5601, t=758.1004, r=324.0271, b=731.8085532416503, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 19))],
 'self_ref': '#/texts/97',
 'source': [],
 'text': 'Nature Alberta News'}

#/texts/98 -----------
{'children': [],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'formatting': None,
 'hyperlink': None,
 'label': <DocItemLabel.TEXT: 'text'>,
 'meta': None,
 'orig': 'JOHANE JANELLE',
 'parent': RefItem(cref='#/pictures/11'),
 'prov': [ProvenanceItem(page_no=5, bbox=BoundingBox(l=327.2545, t=554.1897, r=378.0073, b=548.0050022794846, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0

**Comments:**

From the analysis of this pdf (page 5 selected), I found that:
- Not only 'DocItemLabel.TEXT' are extracted correctly, but 'DocItemLabel.SECTION_HEADER', 'DocItemLabel.PAGE_FOOTER', etc also extracted correctly.
- This means in the previous pdf there is some problem or some odd structure might be used to create the pdf.
- Need to inspect that PDF what's wrong, but first I'll try to understand how images are extracted by docling. 

### Experimenting on pictures/images

In [59]:
data_folder_path = Path("/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named")
file_path = data_folder_path / "nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf"
file_path

PosixPath('/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named/nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf')

In [60]:
converter = DocumentConverter()
doc = converter.convert(file_path).document

[INFO] 2026-04-15 16:13:40,724 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 16:13:40,841 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 16:13:40,843 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 16:13:41,094 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 16:13:41,122 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer.onnx
[INFO] 2026-04-15 16:13:41,123 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer

In [62]:
# doc.export_to_markdown()
# type(doc)
docdict = dict(doc)
docdict

{'schema_name': 'DoclingDocument',
 'version': '1.10.0',
 'name': 'nature-alberta-spring-2026-magazine-medium-mix-multi-col-text',
 'origin': DocumentOrigin(mimetype='application/pdf', binary_hash=9295684724258981057, filename='nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf', uri=None),
 'furniture': GroupItem(self_ref='#/furniture', parent=None, children=[], content_layer=<ContentLayer.FURNITURE: 'furniture'>, meta=None, name='_root_', label=<GroupLabel.UNSPECIFIED: 'unspecified'>),
 'body': GroupItem(self_ref='#/body', parent=None, children=[RefItem(cref='#/texts/0'), RefItem(cref='#/pictures/0'), RefItem(cref='#/texts/5'), RefItem(cref='#/texts/6'), RefItem(cref='#/texts/7'), RefItem(cref='#/texts/8'), RefItem(cref='#/texts/9'), RefItem(cref='#/texts/10'), RefItem(cref='#/texts/11'), RefItem(cref='#/pictures/1'), RefItem(cref='#/texts/14'), RefItem(cref='#/texts/15'), RefItem(cref='#/texts/16'), RefItem(cref='#/texts/17'), RefItem(cref='#/pictures/2'), RefItem(cre

In [63]:
docdict.keys()

dict_keys(['schema_name', 'version', 'name', 'origin', 'furniture', 'body', 'groups', 'texts', 'pictures', 'tables', 'key_value_items', 'form_items', 'field_regions', 'field_items', 'pages'])

In [64]:
type(docdict['pages']), docdict['pages']

(dict,
 {1: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=1),
  2: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=2),
  3: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=3),
  4: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=4),
  5: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5),
  6: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=6),
  7: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=7),
  8: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=8),
  9: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=9),
  10: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=10),
  11: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=11),
  12: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=12),
  13: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=13

In [65]:
docdict['pages'][5]

PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5)

In [66]:
docdict['pictures']

[PictureItem(self_ref='#/pictures/0', parent=RefItem(cref='#/body'), children=[RefItem(cref='#/texts/1'), RefItem(cref='#/texts/2'), RefItem(cref='#/texts/3'), RefItem(cref='#/texts/4')], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.PICTURE: 'picture'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=482.4024658203125, t=756.8190689086914, r=594.837890625, b=732.1046447753906, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 0))], source=[], comments=[], captions=[], references=[], footnotes=[], image=None, annotations=[]),
 PictureItem(self_ref='#/pictures/1', parent=RefItem(cref='#/body'), children=[RefItem(cref='#/texts/12'), RefItem(cref='#/texts/13')], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.PICTURE: 'picture'>, prov=[ProvenanceItem(page_no=2, bbox=BoundingBox(l=109.3674087524414, t=733.0786170959473, r=229.9933319091797, b=637.7931671142578, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), ch

In [67]:
len(docdict['pictures']), type(docdict['pictures'])

(82, list)

In [68]:
dict(docdict['pictures'][0])

{'self_ref': '#/pictures/0',
 'parent': RefItem(cref='#/body'),
 'children': [RefItem(cref='#/texts/1'),
  RefItem(cref='#/texts/2'),
  RefItem(cref='#/texts/3'),
  RefItem(cref='#/texts/4')],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'meta': None,
 'label': <DocItemLabel.PICTURE: 'picture'>,
 'prov': [ProvenanceItem(page_no=1, bbox=BoundingBox(l=482.4024658203125, t=756.8190689086914, r=594.837890625, b=732.1046447753906, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 0))],
 'source': [],
 'comments': [],
 'captions': [],
 'references': [],
 'footnotes': [],
 'image': None,
 'annotations': []}

In [71]:
for picitem in docdict['pictures']:
    picitem = dict(picitem)
    if picitem['prov'][0].page_no == 5:
        print(f"{picitem['self_ref']} -----------")
        pprint(picitem)
        print()

#/pictures/11 -----------
{'annotations': [],
 'captions': [],
 'children': [RefItem(cref='#/texts/98')],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'footnotes': [],
 'image': None,
 'label': <DocItemLabel.PICTURE: 'picture'>,
 'meta': None,
 'parent': RefItem(cref='#/body'),
 'prov': [ProvenanceItem(page_no=5, bbox=BoundingBox(l=52.973358154296875, t=704.6543884277344, r=383.9338073730469, b=543.9897613525391, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 0))],
 'references': [],
 'self_ref': '#/pictures/11',
 'source': []}

#/pictures/12 -----------
{'annotations': [],
 'captions': [],
 'children': [RefItem(cref='#/texts/125'),
              RefItem(cref='#/texts/126'),
              RefItem(cref='#/texts/127'),
              RefItem(cref='#/texts/128')],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'footnotes': [],
 'image': None,
 'label': <DocItemLabel.PICTURE: 'picture'>,
 'meta': None,
 'parent': RefItem(cref='#/body

**Comments:**
- Only bounding box is given for images, but images are not given
- How to get / save images?
- Here is the Docling documentation page for DocumentConverter: [doc page](https://docling-project.github.io/docling/reference/document_converter/)
- 

In [74]:
from docling.document_converter import DocumentConverter
from docling.datamodel.base_models import InputFormat
from docling.datamodel.pipeline_options import PdfPipelineOptions
from docling.document_converter import PdfFormatOption

**1. Enable image retention in PDF pipeline**

By default, images may not be preserved unless configured.

In [77]:
data_folder_path = Path("/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named")
file_path = data_folder_path / "nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf"
file_path

PosixPath('/mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/iid-platform/sample_data/named/nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf')

In [78]:
converter = DocumentConverter(
    format_options={
        InputFormat.PDF: PdfFormatOption(
            pipeline_options=PdfPipelineOptions(
                keep_images=True   # VERY IMPORTANT
            )
        )
    }
)

This uses the PDF pipeline (StandardPdfPipeline) which supports image retention.

**2. Convert the PDF**

In [79]:
result = converter.convert(file_path)
doc = result.document

[INFO] 2026-04-15 19:00:27,086 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 19:00:27,635 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 19:00:27,636 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_PP-OCRv4_det_infer.onnx
[INFO] 2026-04-15 19:00:28,155 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-04-15 19:00:28,233 [RapidOCR] download_file.py:60: File exists and is valid: /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer.onnx
[INFO] 2026-04-15 19:00:28,234 [RapidOCR] main.py:53: Using /mnt/c/Users/Rakesh-PC/Documents/1_GitHubSync_SSH/.llmenv/lib/python3.10/site-packages/rapidocr/models/ch_ppocr_mobile_v2.0_cls_infer

In [80]:
type(doc)

docling_core.types.doc.document.DoclingDocument

In [81]:
docdict = dict(doc)
docdict

{'schema_name': 'DoclingDocument',
 'version': '1.10.0',
 'name': 'nature-alberta-spring-2026-magazine-medium-mix-multi-col-text',
 'origin': DocumentOrigin(mimetype='application/pdf', binary_hash=9295684724258981057, filename='nature-alberta-spring-2026-magazine-medium-mix-multi-col-text.pdf', uri=None),
 'furniture': GroupItem(self_ref='#/furniture', parent=None, children=[], content_layer=<ContentLayer.FURNITURE: 'furniture'>, meta=None, name='_root_', label=<GroupLabel.UNSPECIFIED: 'unspecified'>),
 'body': GroupItem(self_ref='#/body', parent=None, children=[RefItem(cref='#/texts/0'), RefItem(cref='#/pictures/0'), RefItem(cref='#/texts/5'), RefItem(cref='#/texts/6'), RefItem(cref='#/texts/7'), RefItem(cref='#/texts/8'), RefItem(cref='#/texts/9'), RefItem(cref='#/texts/10'), RefItem(cref='#/texts/11'), RefItem(cref='#/pictures/1'), RefItem(cref='#/texts/14'), RefItem(cref='#/texts/15'), RefItem(cref='#/texts/16'), RefItem(cref='#/texts/17'), RefItem(cref='#/pictures/2'), RefItem(cre

In [82]:
docdict.keys()

dict_keys(['schema_name', 'version', 'name', 'origin', 'furniture', 'body', 'groups', 'texts', 'pictures', 'tables', 'key_value_items', 'form_items', 'field_regions', 'field_items', 'pages'])

In [83]:
type(docdict['pages']), docdict['pages']

(dict,
 {1: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=1),
  2: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=2),
  3: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=3),
  4: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=4),
  5: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5),
  6: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=6),
  7: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=7),
  8: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=8),
  9: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=9),
  10: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=10),
  11: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=11),
  12: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=12),
  13: PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=13

In [84]:
docdict['pages'][5]

PageItem(size=Size(width=612.0, height=792.0), image=None, page_no=5)

In [85]:
docdict['pictures']

[PictureItem(self_ref='#/pictures/0', parent=RefItem(cref='#/body'), children=[RefItem(cref='#/texts/1'), RefItem(cref='#/texts/2'), RefItem(cref='#/texts/3'), RefItem(cref='#/texts/4')], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.PICTURE: 'picture'>, prov=[ProvenanceItem(page_no=1, bbox=BoundingBox(l=482.4024658203125, t=756.8190689086914, r=594.837890625, b=732.1046447753906, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 0))], source=[], comments=[], captions=[], references=[], footnotes=[], image=None, annotations=[]),
 PictureItem(self_ref='#/pictures/1', parent=RefItem(cref='#/body'), children=[RefItem(cref='#/texts/12'), RefItem(cref='#/texts/13')], content_layer=<ContentLayer.BODY: 'body'>, meta=None, label=<DocItemLabel.PICTURE: 'picture'>, prov=[ProvenanceItem(page_no=2, bbox=BoundingBox(l=109.3674087524414, t=733.0786170959473, r=229.9933319091797, b=637.7931671142578, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), ch

In [86]:
len(docdict['pictures']), type(docdict['pictures'])

(82, list)

In [87]:
dict(docdict['pictures'][0])

{'self_ref': '#/pictures/0',
 'parent': RefItem(cref='#/body'),
 'children': [RefItem(cref='#/texts/1'),
  RefItem(cref='#/texts/2'),
  RefItem(cref='#/texts/3'),
  RefItem(cref='#/texts/4')],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'meta': None,
 'label': <DocItemLabel.PICTURE: 'picture'>,
 'prov': [ProvenanceItem(page_no=1, bbox=BoundingBox(l=482.4024658203125, t=756.8190689086914, r=594.837890625, b=732.1046447753906, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 0))],
 'source': [],
 'comments': [],
 'captions': [],
 'references': [],
 'footnotes': [],
 'image': None,
 'annotations': []}

In [88]:
for picitem in docdict['pictures']:
    picitem = dict(picitem)
    if picitem['prov'][0].page_no == 5:
        print(f"{picitem['self_ref']} -----------")
        pprint(picitem)
        print()

#/pictures/11 -----------
{'annotations': [],
 'captions': [],
 'children': [RefItem(cref='#/texts/98')],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'footnotes': [],
 'image': None,
 'label': <DocItemLabel.PICTURE: 'picture'>,
 'meta': None,
 'parent': RefItem(cref='#/body'),
 'prov': [ProvenanceItem(page_no=5, bbox=BoundingBox(l=52.973358154296875, t=704.6543884277344, r=383.9338073730469, b=543.9897613525391, coord_origin=<CoordOrigin.BOTTOMLEFT: 'BOTTOMLEFT'>), charspan=(0, 0))],
 'references': [],
 'self_ref': '#/pictures/11',
 'source': []}

#/pictures/12 -----------
{'annotations': [],
 'captions': [],
 'children': [RefItem(cref='#/texts/125'),
              RefItem(cref='#/texts/126'),
              RefItem(cref='#/texts/127'),
              RefItem(cref='#/texts/128')],
 'comments': [],
 'content_layer': <ContentLayer.BODY: 'body'>,
 'footnotes': [],
 'image': None,
 'label': <DocItemLabel.PICTURE: 'picture'>,
 'meta': None,
 'parent': RefItem(cref='#/body

Key observation from the official docs:

**What is mentioned?**
- DocumentConverter → converts documents
- StandardPdfPipeline → processing pipeline
- Output → DoclingDocument (structured)

**What is NOT mentioned anywhere**
- ❌ No API to render page regions
- ❌ No API to export images from bbox
- ❌ No guarantee that pictures[i].image is populated
- ❌ No method like get_image() / render_region()

👉 That omission is intentional.

```
          Docling
            ↓
   structured layout (texts, pictures, bbox)
            ↓
     + external renderer
            ↓
   cropped images aligned with structure
```

1. Open PDF once (important for performance)

In [92]:
import fitz  # PyMuPDF

pdf = fitz.open(file_path)

2. Function to crop any Docling picture

In [93]:
def render_picture(picitem, pdf, zoom=2):
    # Get page (Docling is 1-based, PyMuPDF is 0-based)
    page_no = picitem['prov'][0].page_no - 1
    page = pdf[page_no]

    bbox = picitem['prov'][0].bbox

    # Create rectangle (Docling uses bottom-left origin → compatible)
    rect = fitz.Rect(bbox.l, bbox.b, bbox.r, bbox.t)

    # Increase resolution (important for RAG quality)
    mat = fitz.Matrix(zoom, zoom)

    pix = page.get_pixmap(matrix=mat, clip=rect)

    return pix

In [95]:
def render_picture(picitem, pdf, zoom=2):
    page_no = picitem['prov'][0].page_no - 1
    page = pdf[page_no]

    bbox = picitem['prov'][0].bbox
    page_height = page.rect.height

    # Correct coordinate transform
    x0 = bbox.l
    x1 = bbox.r
    y0 = page_height - bbox.t
    y1 = page_height - bbox.b

    rect = fitz.Rect(x0, y0, x1, y1)

    mat = fitz.Matrix(zoom, zoom)
    pix = page.get_pixmap(matrix=mat, clip=rect)

    return pix

3. Save or view the image

In [96]:
for picitem in docdict['pictures']:
    picitem = dict(picitem)

    if picitem['prov'][0].page_no == 5:
        pix = render_picture(picitem, pdf)

        out_path = f"{picitem['self_ref'].split('/')[-1]}.png"
        pix.save(out_path)

        print(f"Saved: {out_path}")

Saved: 11.png
Saved: 12.png
